manual/tools/caml_tex2.ml

```diff
@@ -64,6 +53,164 @@ let (~!) =
       memo := (key, data) :: !memo;
       data
 
+exception Phrase_parsing of string
+
+module Toplevel = struct
+  (** Initialize the toplevel loop, redirect stdout and stderr,
+      capture warnings and error messages *)
+
+  type output =
+    {
+      error : string; (** error message text *)
+      warnings : string list; (** warning messages text *)
+      values : string; (** toplevel output *)
+      stdout : string; (** output printed on the toplevel stdout *)
+      underlined : (int * int) list
+      (** locations to underline in input phrases *)
+    }
+
+  let buffer_fmt () =
+    let b = Buffer.create 30 in b, Format.formatter_of_buffer b
+
+  let error_fmt = buffer_fmt ()
+  let warning_fmt = buffer_fmt ()
+
+  let out_fmt = buffer_fmt ()
+
+  let flush_fmt (b,fmt) =
+    Format.pp_print_flush fmt ();
+    let r = Buffer.contents b in
+    Buffer.reset b;
+    r
+
+  (** Redirect the stdout *)
+  let stdout_out, stdout_in = Unix.pipe ~cloexec:true ()
+  let () = Unix.dup2 stdout_in Unix.stdout;
+    Unix.set_nonblock stdout_out
+
+  let self_error_fmt = Format.formatter_of_out_channel stderr
+  let eprintf = Format.eprintf
+
+  let read_stdout =
+    let size = 50 in
+    let b = Bytes.create size in
+    let buffer = Buffer.create 100 in
+    let max_retry = 5 in
+    let rec loop retry =
+      try
+        let n = Unix.read stdout_out b 0 size in
+        if n = size then
+          (Buffer.add_bytes buffer b; loop max_retry)
+        else
+          Buffer.add_bytes buffer (Bytes.sub b 0 n)
```

Comment:
This could be simpler with an un-conditional `Buffer.add_subbytes buffer b 0 n`.

Reply:
Fixed.